system-addon/lib/HighlightsFeed.jsm

```diff
@@ -55,11 +59,18 @@ this.HighlightsFeed = class HighlightsFeed {
     SectionsManager.disableSection(SECTION_ID);
     PageThumbs.removeExpirationFilter(this);
     Services.obs.removeObserver(this, SYNC_BOOKMARKS_FINISHED_EVENT);
+    Services.obs.removeObserver(this, BOOKMARKS_RESTORE_SUCCESS_EVENT);
+    Services.obs.removeObserver(this, BOOKMARKS_RESTORE_FAILED_EVENT);
   }
 
   observe(subject, topic, data) {
-    // When we receive a notification that a sync has happened for bookmarks, refresh highlights
-    if (topic === SYNC_BOOKMARKS_FINISHED_EVENT && data === "bookmarks") {
+    // When we receive a notification that a sync has happened for bookmarks,
+    // or Places finished importing or restoring bookmarks, refresh highlights
+    const manyBookmarksChanged =
+      (topic === SYNC_BOOKMARKS_FINISHED_EVENT && data === "bookmarks") ||
+      topic === BOOKMARKS_RESTORE_SUCCESS_EVENT ||
+      topic === BOOKMARKS_RESTORE_FAILED_EVENT;
```

Comment:
I'm not sure we actually _do_ want to do a refresh if an restore has failed. Maybe you can give some more context about when this event is fired, and why we would want to fetch new highlights in this case?

Reply:
That makes sense. If a restore fails, it'll erase all your existing bookmarks, but not add any new ones...so I guess the only issue is that existing highlights will be stale. But, Places tries its best to validate and fix up malformed backups before erasing everything, so do you think it's worth removing `BOOKMARKS_RESTORE_FAILED_EVENT`?

Reply:
Does the event still fire even if Places manages to fix things up and not erase everything?

Reply:
Nope, it'll be `SUCCESS` in that case.

Reply:
Ok then I think it's ok to leave the failed even. If it really truly does fail and the bookmarks are all gone then we should reflect that in Activity Stream